src/client-side-metrics/operation-metrics-collector.ts

```diff
@@ -0,0 +1,337 @@
+   */
+  onAttemptComplete(projectId: string, info: OnAttemptCompleteInfo) {
+    if (
+      this.state ===
+        MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_NO_ROWS_YET ||
+      this.state ===
+        MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_SOME_ROWS_RECEIVED
+    ) {
+      this.state =
+        MetricsCollectorState.OPERATION_STARTED_ATTEMPT_NOT_IN_PROGRESS;
+      this.attemptCount++;
+      const endTime = new Date();
+      if (projectId && this.attemptStartTime) {
+        const attributes = {
+          streamingOperation: info.streamingOperation,
+          attemptStatus: info.attemptStatus,
+          clientName: `nodejs-bigtable/${version}`,
+          metricsCollectorData: this.getMetricsCollectorData(),
+          projectId,
+        };
+        const totalTime = endTime.getTime() - this.attemptStartTime.getTime();
+        this.metricsHandlers.forEach(metricsHandler => {
+          if (metricsHandler.onAttemptComplete) {
+            metricsHandler.onAttemptComplete(
+              {
+                attemptLatency: totalTime,
+                serverLatency: this.serverTime ?? undefined,
+                connectivityErrorCount: this.connectivityErrorCount,
+              },
+              attributes
+            );
+          }
+        });
+      }
+    } else {
+      console.warn('Invalid state transition attempted');
+    }
+  }
+
+  /**
+   * Called when a new attempt starts. Records the start time of the attempt.
+   */
+  onAttemptStart() {
+    if (
+      this.state ===
+      MetricsCollectorState.OPERATION_STARTED_ATTEMPT_NOT_IN_PROGRESS
+    ) {
+      this.state =
+        MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_NO_ROWS_YET;
+      this.attemptStartTime = new Date();
+      this.serverTime = null;
+      this.serverTimeRead = false;
+      this.connectivityErrorCount = 0;
+    } else {
+      console.warn('Invalid state transition attempted');
+    }
+  }
+
+  /**
+   * Called when the first response is received. Records first response latencies.
+   */
+  onResponse(projectId: string) {
+    if (
+      this.state ===
+      MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_NO_ROWS_YET
+    ) {
+      this.state =
+        MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_SOME_ROWS_RECEIVED;
+      const endTime = new Date();
+      if (projectId && this.operationStartTime) {
+        this.firstResponseLatency =
+          endTime.getTime() - this.operationStartTime.getTime();
+      }
+    }
+  }
+
+  /**
+   * Called when an operation completes (successfully or unsuccessfully).
+   * Records operation latencies, retry counts, and connectivity error counts.
+   * @param {string} projectId The id of the project.
+   * @param {OperationOnlyAttributes} info Information about the completed operation.
+   */
+  onOperationComplete(projectId: string, info: OperationOnlyAttributes) {
```

Comment:
Same question as onAttemptComplete: don't we just need the status code here?

Reply:
We could definitely pass in the _streamingOperation_ parameter when we create the metrics collector instead so we don't need to pass it in here. _projectId_ is a different story, because the projectId property on the client is only populated by calling auth just before the grpc call is made so it's not available when the metrics collector gets created.

Reply:
Ok _streamingOperation_ has been removed.